### PR TITLE
Add dummy POCOs for Bash, PowerShell and Publish steps

### PR DESCRIPTION
**Pull request by premun**

### CODE REVIEW

**Comment by mandel-macaque** [comment]
I want to wait to do the review based on what we decide regarding #9

**Comment by premun** [comment]
@mandel-macaque I have moved the optional parameters out of ctors. Required are usually just names and display names.

We could check this in if you agree and I would start building the API.

File: src/Sharpliner.Model/Job.cs [hunk 1]
```diff
@@ -3,20 +3,40 @@
...
-        IEnumerable<Step> Steps,
-        JobWorkspace Workspace = JobWorkspace.Outputs,
-        bool ContinueOnError = false);
 
     // TODO: missing properties Uses, Services
+    public record Job
+    {
+        public string Name { get; }
+
+        public string DisplayName { get; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Job.cs:12]: Nullable? That will help us to validate.

```suggestion
        public string? DisplayName { get; }
```

File: src/Sharpliner.Model/Job.cs [hunk 2]
```diff
@@ -3,20 +3,40 @@
...
+
+        public ContainerReference? Container { get; init; }
+
+        public TimeSpan? Timeout { get; init; }
+
+        public TimeSpan? CancelTimeout { get; init; }
+
+        public IEnumerable<Variable>? Variables { get; init; }
+
+        public IEnumerable<Step>? Steps { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Job.cs:30]: This is going to have to be a list and be initialize already. That way users just have to use {} which results in a call to .Add
```suggestion
        public List<Step> Steps { get; } = new List<Step>();
```
If you try the API as you wrote it the user will have to do:
```csharp 
Steps = new [] {}
// or
Steps = new List<Step> ();
// But if the list is already present
Steps = {
new Step()
}
```

File: src/Sharpliner.Model/Job.cs [hunk 3]
```diff
@@ -3,20 +3,40 @@
...
+
+        public JobWorkspace Workspace { get; init; } = JobWorkspace.Outputs;
+
+        public bool ContinueOnError { get; init; } = false;
+
+        public Job(string name, string displayName)
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Job.cs:38]: So we need that with nullable? I mean, with and older version of C# the API is horrible...

File: src/Sharpliner.Model/Pipeline.cs [hunk 4]
```diff
@@ -2,10 +2,16 @@
...
-        IEnumerable<Variable> Variables,
-        IEnumerable<Trigger> Triggers,
-        IEnumerable<Stage> Stages);
+    public record Pipeline
+    {
+        public string? Name { get; init; }
+
+        public Resources? Resources { get; init; }
+
+        public IEnumerable<Variable>? Variables { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Pipeline.cs:11]: Same, List and init it

File: src/Sharpliner.Model/Pipeline.cs [hunk 5]
```diff
@@ -2,10 +2,16 @@
...
+
+        public IEnumerable<Trigger>? Triggers { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Pipeline.cs:13]: Same

File: src/Sharpliner.Model/Pipeline.cs [hunk 6]
```diff
@@ -2,10 +2,16 @@
...
+
+        public IEnumerable<Stage>? Stages { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Pipeline.cs:15]: Same

File: src/Sharpliner.Model/Pool.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        internal PoolName(string name) : base()
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
+        }
+    }
+
+    public record HostedPool(
+        string Name,
+        string VmImage,
+        IEnumerable<string>? Demands = null)
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Pool.cs:21]: Move demands to a property too

File: src/Sharpliner.Model/Stage.cs [hunk 8]
```diff
@@ -2,11 +2,24 @@
...
-        string Condition,
-        IEnumerable<Variable> Variables,
-        IEnumerable<Job> Jobs);
+    public record Stage
+    {
+        public string Name { get; }
+
+        public string DisplayName { get; }
+
+        public IEnumerable<string>? DependsOn { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Stage.cs:11]: Same comment about the list

File: src/Sharpliner.Model/Stage.cs [hunk 9]
```diff
@@ -2,11 +2,24 @@
...
+
+        public string? Condition { get; init; }
+
+        public IEnumerable<Variable>? Variables { get; init; }
```
Comment (mandel-macaque) [inline src/Sharpliner.Model/Stage.cs:15]: Same

**Review by mandel-macaque — CHANGES_REQUESTED** [review]
Small changes about the collections.   made the same mistake on my PR and I noticed because I wrote the test and saw the extra tying. I know it looks stupid, but can you write some dummy tests (ex: throws NRE). They will
Increase coverage, but best of all, we can see the API in use.

File: src/Sharpliner.Model/Job.cs (hunk 2 above)
Comment (premun) [reply src/Sharpliner.Model/Job.cs:30]: Oh my god! This is amazing! I didn't know about this! I was always wondering how we were gonna achieve `{}` only

File: src/Sharpliner.Model/Job.cs (hunk 3 above)
Comment (premun) [reply src/Sharpliner.Model/Job.cs:38]: If someone who doesn't use nullable uses our library, he can pass null in. Or someone else could make a mistake and pass null even when they have nullability on.

I am not 100% sure display name is actually needed by AzDO but maybe we should require it?

**Comment by premun** [comment]
@mandel-macaque I am moving these to the AzDO namespace in the API branch

File: src/Sharpliner.Model/Job.cs (hunk 1 above)
Comment (premun) [reply src/Sharpliner.Model/Job.cs:12]: I don't understand?

**Review by mandel-macaque — APPROVED** [review]

**Pull request merged**